diff -ur /tmp/tt-rss/classes/feedparser.php tt-rss/classes/feedparser.php
--- /tmp/tt-rss/classes/feedparser.php	2013-07-23 12:19:55.000000000 +0800
+++ tt-rss/classes/feedparser.php	2017-02-02 21:40:26.382627084 +0800
@@ -16,9 +16,10 @@
 		libxml_use_internal_errors(true);
 		libxml_clear_errors();
 		$this->doc = new DOMDocument();
-		$pattern = '/^(<\?xml[^<>]*encoding=")([^"]*)"/';
-		if (preg_match($pattern, $data, $matches)) {
-			$enc = strtolower($matches[2]);
+		$pattern = '/(<\?xml[^<>]*encoding=")([^"]*)"/';
+		if (preg_match($pattern, $data, $matches, PREG_OFFSET_CAPTURE)) {
+			$data = substr($data, $matches[0][1]);
+			$enc = strtolower($matches[2][0]);
 			if ($enc != "utf-8") {
 				if ($enc == "gb2312" or $enc == "gbk") {
 					$enc = "cp936";
@@ -26,6 +27,10 @@
 				$data = preg_replace($pattern, '${1}utf-8"', iconv( $enc, 'utf-8', $data));
 			}
 		}
+		$pattern = '/xmlns="(http:..purl.org.atom.ns)([^"]*)"/';
+		if (preg_match($pattern, $data, $matches, PREG_OFFSET_CAPTURE)) {
+			$data = preg_replace($pattern, 'xmlns="http://www.w3.org/2005/Atom"', $data);
+		}
 		$this->doc->loadXML($data);
 		$this->error = $this->format_error(libxml_get_last_error());
 		libxml_clear_errors();
@@ -35,6 +40,9 @@
 
 	function init() {
 		$root = $this->doc->firstChild;
+		if (!$root) {
+			return;
+		}
 		$xpath = new DOMXPath($this->doc);
 		$xpath->registerNamespace('atom', 'http://www.w3.org/2005/Atom');
 		$xpath->registerNamespace('media', 'http://search.yahoo.com/mrss/');
@@ -59,7 +67,7 @@
 				$this->type = $this::FEED_ATOM;
 				break;
 			default:
-				$this->error = "Unknown/unsupported feed type";
+				$this->error = "Unknown/unsupported feed type: " . $root->tagName;
 				return;
 			}
 
@@ -131,7 +139,7 @@
 
 			}
 		} else {
-			$this->error = "Unknown/unsupported feed type";
+			$this->error = "Unknown/unsupported feed type: no root: " . $this->doc->saveXML();
 			return;
 		}
 	}
diff -ur /tmp/tt-rss/classes/pref/feeds.php tt-rss/classes/pref/feeds.php
--- /tmp/tt-rss/classes/pref/feeds.php	2013-05-13 17:34:26.000000000 +0800
+++ tt-rss/classes/pref/feeds.php	2017-05-05 01:12:12.746322929 +0800
@@ -1270,11 +1270,15 @@
 			$interval_qpart = "DATE_SUB(NOW(), INTERVAL 3 MONTH)";
 		}
 
+		/*
 		$result = $this->dbh->query("SELECT COUNT(*) AS num_inactive FROM ttrss_feeds WHERE
 					(SELECT MAX(updated) FROM ttrss_entries, ttrss_user_entries WHERE
 						ttrss_entries.id = ref_id AND
 							ttrss_user_entries.feed_id = ttrss_feeds.id) < $interval_qpart AND
 			ttrss_feeds.owner_uid = ".$_SESSION["uid"]);
+		 */
+		$result = $this->dbh->query("select count(*) as num_inactive from ttrss_feeds left join (select distinct feed_id from ttrss_user_entries inner join (select id FROM ttrss_entries
+			where updated >= $interval_qpart) as dd on ref_id = dd.id) as recent_feeds on id = feed_id where feed_id is null and owner_uid = ".$_SESSION["uid"]);
 
 		$num_inactive = $this->dbh->fetch_result($result, 0, "num_inactive");
 
@@ -1713,8 +1717,20 @@
 			WHERE feed_id IS NULL AND marked = false AND owner_uid = " . $_SESSION["uid"]);
 		}
 
+		/*
 		$result = $this->dbh->query("DELETE FROM ttrss_entries WHERE
 			(SELECT COUNT(int_id) FROM ttrss_user_entries WHERE ref_id = id) = 0");
+		 */
+		$this->dbh->query("create temporary table orphan_ids (primary key pkey(id)) select ttrss_entries.id from ttrss_entries left join ttrss_user_entries
+			on ttrss_entries.id = ttrss_user_entries.ref_id where ttrss_user_entries.ref_id is null");
+		$result = $this->dbh->query("select count(id) as num_ids from orphan_ids");
+		$num_ids = $this->dbh->fetch_result($result, 0, "num_ids");
+		if ($num_ids > 0) {
+			$result = $this->dbh->query("delete from ttrss_entries where id in (select id from orphan_ids);");
+		} else {
+			$result = $this->dbh->query("delete from ttrss_entries where id is null;");
+		}
+		$this->dbh->query("drop temporary table orphan_ids;");
 
 		ccache_update($id, $_SESSION['uid']);
 	} // function clear_feed_articles
diff -ur /tmp/tt-rss/include/functions.php tt-rss/include/functions.php
--- /tmp/tt-rss/include/functions.php	2014-11-22 11:23:59.000000000 +0800
+++ tt-rss/include/functions.php	2017-05-06 10:09:21.604962529 +0800
@@ -276,13 +276,25 @@
 	function purge_orphans($do_output = false) {
 
 		// purge orphaned posts in main content table
+		/*
 		$result = db_query("DELETE FROM ttrss_entries WHERE
 			(SELECT COUNT(int_id) FROM ttrss_user_entries WHERE ref_id = id) = 0");
+		 */
+		db_query("create temporary table orphan_ids (primary key pkey(id)) select ttrss_entries.id from ttrss_entries left join ttrss_user_entries
+			on ttrss_entries.id = ttrss_user_entries.ref_id where ttrss_user_entries.ref_id is null");
+		$result = db_query("select count(id) as num_ids from orphan_ids");
+		$num_ids = db_fetch_result($result, 0, "num_ids");
+		if ($num_ids > 0) {
+			$result = db_query("delete from ttrss_entries where id in (select id from orphan_ids);");
+		} else {
+			$result = db_query("delete from ttrss_entries where id is null;");
+		}
 
 		if ($do_output) {
 			$rows = db_affected_rows($result);
 			_debug("Purged $rows orphaned posts.");
 		}
+		db_query("drop temporary table orphan_ids;");
 	}
 
 	function get_feed_update_interval($feed_id) {
@@ -3739,8 +3751,12 @@
 	}
 
 	function getLastArticleId() {
+		/*
 		$result = db_query("SELECT MAX(ref_id) AS id FROM ttrss_user_entries
 			WHERE owner_uid = " . $_SESSION["uid"]);
+		 */
+		$result = db_query("SELECT ref_id AS id FROM ttrss_user_entries
+			WHERE owner_uid = " . $_SESSION["uid"] . " order by ref_id desc limit 1");
 
 		if (db_num_rows($result) == 1) {
 			return db_fetch_result($result, 0, "id");
diff -ur /tmp/tt-rss/schema/ttrss_schema_mysql.sql tt-rss/schema/ttrss_schema_mysql.sql
--- /tmp/tt-rss/schema/ttrss_schema_mysql.sql	2013-05-13 17:34:26.000000000 +0800
+++ tt-rss/schema/ttrss_schema_mysql.sql	2017-05-06 12:18:56.207502376 +0800
@@ -199,6 +199,7 @@
 create index ttrss_user_entries_ref_id_index on ttrss_user_entries(ref_id);
 create index ttrss_user_entries_feed_id on ttrss_user_entries(feed_id);
 create index ttrss_user_entries_unread_idx on ttrss_user_entries(unread);
+create index ttrss_user_entries_marked_unread_idx on ttrss_user_entries(marked, unread);
 
 create table ttrss_entry_comments (id integer not null primary key,
 	ref_id integer not null,
