--- /tmp/classes/feedparser.php	2013-05-13 17:34:26.000000000 +0800
+++ ./classes/feedparser.php	2013-07-23 12:19:55.781343698 +0800
@@ -16,6 +16,16 @@
 		libxml_use_internal_errors(true);
 		libxml_clear_errors();
 		$this->doc = new DOMDocument();
+		$pattern = '/^(<\?xml[^<>]*encoding=")([^"]*)"/';
+		if (preg_match($pattern, $data, $matches)) {
+			$enc = strtolower($matches[2]);
+			if ($enc != "utf-8") {
+				if ($enc == "gb2312" or $enc == "gbk") {
+					$enc = "cp936";
+				}
+				$data = preg_replace($pattern, '${1}utf-8"', iconv( $enc, 'utf-8', $data));
+			}
+		}
 		$this->doc->loadXML($data);
 		$this->error = $this->format_error(libxml_get_last_error());
 		libxml_clear_errors();
