diff --git a/src/Makefile.am b/src/Makefile.am
index b926e7d..8b2e5cf 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -6,6 +6,7 @@ if !USE_SYSTEM_SHARED_LIB
 AM_CFLAGS += -I$(top_srcdir)/libbloom
 AM_CFLAGS += -I$(top_srcdir)/libipset/include
 AM_CFLAGS += -I$(top_srcdir)/libcork/include
+AM_CFLAGS += -I$(top_srcdir)/../libancillary
 endif
 AM_CFLAGS += $(LIBPCRE_CFLAGS)
 
@@ -13,11 +14,13 @@ SS_COMMON_LIBS = $(INET_NTOP_LIB) $(LIBPCRE_LIBS)
 if !USE_SYSTEM_SHARED_LIB
 SS_COMMON_LIBS += $(top_builddir)/libbloom/libbloom.la \
                   $(top_builddir)/libipset/libipset.la \
+                  $(top_builddir)/../libancillary/libancillary.a \
                   $(top_builddir)/libcork/libcork.la
 else
 SS_COMMON_LIBS += -lbloom -lcork -lcorkipset
 endif
 SS_COMMON_LIBS += -lev -lsodium -lm
+SS_COMMON_LIBS += -llog
 
 bin_PROGRAMS = ss-local ss-tunnel ss-server
 if !BUILD_WINCOMPAT
@@ -43,6 +46,7 @@ common_src = utils.c \
              json.c \
              udprelay.c \
              cache.c \
+	     android.c \
              netutils.c
 
 if BUILD_WINCOMPAT
@@ -72,6 +76,7 @@ ss_server_SOURCES = resolv.c \
 ss_manager_SOURCES = utils.c \
                      jconf.c \
                      json.c \
+		     android.c \
                      netutils.c \
                      manager.c
 
@@ -95,6 +100,7 @@ bin_PROGRAMS += ss-redir
 ss_redir_SOURCES = utils.c \
                    jconf.c \
                    json.c \
+		   android.c \
                    netutils.c \
                    cache.c \
                    udprelay.c \
diff --git a/src/manager.c b/src/manager.c
index 742cdf6..58a9ad8 100644
--- a/src/manager.c
+++ b/src/manager.c
@@ -65,6 +65,11 @@
 #define BUF_SIZE 65535
 #endif
 
+#ifdef __ANDROID__
+char *stat_path   = NULL;
+int vpn = 0;
+#endif
+
 int verbose          = 0;
 char *executable     = "ss-server";
 char *working_dir    = NULL;
diff --git a/src/redir.c b/src/redir.c
index 42dafb4..5685aef 100644
--- a/src/redir.c
+++ b/src/redir.c
@@ -84,6 +84,11 @@ int reuse_port = 0;
 
 static crypto_t *crypto;
 
+#ifdef __ANDROID__
+char *stat_path   = NULL;
+int vpn = 0;
+#endif
+
 static int ipv6first = 0;
 static int mode      = TCP_ONLY;
 #ifdef HAVE_SETRLIMIT
diff --git a/src/server.c b/src/server.c
index a686743..96560eb 100644
--- a/src/server.c
+++ b/src/server.c
@@ -115,6 +115,11 @@ struct sockaddr_storage local_addr_v6;
 
 static crypto_t *crypto;
 
+#ifdef __ANDROID__
+char *stat_path   = NULL;
+int vpn = 0;
+#endif
+
 static int acl       = 0;
 static int mode      = TCP_ONLY;
 static int ipv6first = 0;
diff --git a/src/tunnel.c b/src/tunnel.c
index a928e61..2c361ef 100644
--- a/src/tunnel.c
+++ b/src/tunnel.c
@@ -77,6 +77,7 @@ static void free_server(server_t *server);
 static void close_and_free_server(EV_P_ server_t *server);
 
 #ifdef __ANDROID__
+char *stat_path   = NULL;
 int vpn = 0;
 #endif
 
diff --git a/src/utils.h b/src/utils.h
index 450e3ec..f5615fe 100644
--- a/src/utils.h
+++ b/src/utils.h
@@ -37,6 +37,7 @@
 #include <android/log.h>
 #define USE_TTY()
 #define USE_SYSLOG(ident, _cond)
+#define USE_LOGFILE(ident) ((void)(ident))
 #define LOGI(...)                                                \
     ((void)__android_log_print(ANDROID_LOG_DEBUG, "shadowsocks", \
                                __VA_ARGS__))
